src/audio/pipeline.c

```diff
@@ -389,11 +389,62 @@ static int pipeline_comp_hw_params(struct comp_dev *current,
 	return 0;
 }
 
+static int pipeline_comp_params_neg(struct comp_dev *current,
+				    struct comp_buffer *calling_buf,
+				    struct pipeline_walk_context *ctx,
+				    int dir)
+{
+	struct pipeline_data *ppl_data = ctx->comp_data;
+	uint32_t flags = 0;
+	int err = 0;
+
+	pipe_dbg(ppl_data->p, "pipeline_comp_params_neg(), current->comp.id = %u, dir = %u",
+		 dev_comp_id(current), dir);
+
+	/* return if current is already configured */
+	if (current->state != COMP_STATE_INIT &&
+	    current->state != COMP_STATE_READY) {
+		if (buffer_params_match(calling_buf,
+					&ppl_data->params->params))
+			return 0;
+		/*
+		 * the param is conflict with an active pipeline,
+		 * drop an error and reject the .params() command.
+		 */
+		pipe_err(ppl_data->p, "pipeline_comp_params_neg(): params conflict with existed active pipeline!");
+		return -EINVAL;
+	}
+
+	/*
+	 * Negotiation only happen when the current component has > 1
+	 * source or sink, we are propagating the params to branched
+	 * buffers, and the subsequent component's .params() or .prepare()
+	 * should be responsible for calibrating if needed. For example,
+	 * a component who has different channels input/output buffers
+	 * should explicitly configure the channels of the branched buffers.
+	 */
+	if (calling_buf) {
+		buffer_lock(calling_buf, &flags);
+		err = buffer_set_params(calling_buf,
+					&ppl_data->params->params,
+					BUFFER_UPDATE_FORCE);
+		buffer_unlock(calling_buf, flags);
+	}
+
+	/* todo: do we need to go further? */
```

Comment:
Do you mean walk the new pipeline here. I think at this point the connected pipeline source buffer will have the correct params configured so no further action would be needed. This can be removed.

Reply:
Yes, we stop here after the connected (but *not paramed*) pipeline source buffer is propagated.
We might need to propagate it forward in the future, but no this requirement for now.

OK, let me remove this comment.

Reply:
It's still there.